<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Working with haplotypes · SequenceVariation.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="SequenceVariation.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">SequenceVariation.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>Working with haplotypes</a><ul class="internal"><li><a class="tocitem" href="#Calling-variants"><span>Calling variants</span></a></li><li><a class="tocitem" href="#Sequence-reconstruction"><span>Sequence reconstruction</span></a></li><li><a class="tocitem" href="#Reference-switching"><span>Reference switching</span></a></li></ul></li><li><a class="tocitem" href="../variations/">Working with variations</a></li><li><a class="tocitem" href="../compare/">Comparing variations</a></li><li><a class="tocitem" href="../api/">API Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Working with haplotypes</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Working with haplotypes</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/BioJulia/SequenceVariation.jl/blob/master/docs/src/haplotypes.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Working-with-haplotypes"><a class="docs-heading-anchor" href="#Working-with-haplotypes">Working with haplotypes</a><a id="Working-with-haplotypes-1"></a><a class="docs-heading-anchor-permalink" href="#Working-with-haplotypes" title="Permalink"></a></h1><h2 id="Calling-variants"><a class="docs-heading-anchor" href="#Calling-variants">Calling variants</a><a id="Calling-variants-1"></a><a class="docs-heading-anchor-permalink" href="#Calling-variants" title="Permalink"></a></h2><p>The first step in working with sequence variation is to identify (call) variations between two sequences. SequenceVariation can directly call variants using the <code>Haplotype(::PairwiseAlignment)</code> constructor of the <a href="../api/#SequenceVariation.Haplotype"><code>Haplotype</code></a> type.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; using SequenceVariation, BioAlignments, BioSequences</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; bovine = dna&quot;GACCGGCTGCATTCGAGGCTGCCAGCAAGCAG&quot;;</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; ovine  = dna&quot;GACCGGCTGCATTCGAGGCTGTCAGCAAACAG&quot;;</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; human  = dna&quot;GACAGGCTGCATCAGAAGAGGCCATCAAGCAG&quot;;</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; bos_ovis_alignment =
           PairwiseAlignment(AlignedSequence(ovine, Alignment(&quot;32M&quot;, 1, 1)), bovine);</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; bos_human_alignment =
           PairwiseAlignment(AlignedSequence(human, Alignment(&quot;32M&quot;, 1, 1)), bovine);</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; bos_ovis_haplotype = Haplotype(bos_ovis_alignment)</code><code class="nohighlight hljs ansi" style="display:block;">Haplotype{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSymbols.DNA} with 2 edits:
  C22T
  G29A</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; bos_human_haplotype = Haplotype(bos_human_alignment)</code><code class="nohighlight hljs ansi" style="display:block;">Haplotype{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSymbols.DNA} with 7 edits:
  C4A
  T13C
  C14A
  G17A
  C19A
  T20G
  G25T</code></pre><h2 id="Sequence-reconstruction"><a class="docs-heading-anchor" href="#Sequence-reconstruction">Sequence reconstruction</a><a id="Sequence-reconstruction-1"></a><a class="docs-heading-anchor-permalink" href="#Sequence-reconstruction" title="Permalink"></a></h2><p>If the alternate sequence of a haplotype is no longer available (as is often the case when calling variants from alignment files), then the sequence can be retrieved using the <a href="../api/#SequenceVariation.reconstruct"><code>reconstruct</code></a> function.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; human2 = reconstruct(bos_human_haplotype)</code><code class="nohighlight hljs ansi" style="display:block;">32nt DNA Sequence:
GACAGGCTGCATCAGAAGAGGCCATCAAGCAG</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; human2 == bovine</code><code class="nohighlight hljs ansi" style="display:block;">false</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; human2 == human</code><code class="nohighlight hljs ansi" style="display:block;">true</code></pre><h2 id="Reference-switching"><a class="docs-heading-anchor" href="#Reference-switching">Reference switching</a><a id="Reference-switching-1"></a><a class="docs-heading-anchor-permalink" href="#Reference-switching" title="Permalink"></a></h2><p>All variations within a haplotype can be mapped to a new reference sequence given an alignment between the new and old references using the <a href="../api/#SequenceVariation.translate-Union{Tuple{T}, Tuple{S}, Tuple{Haplotype{S, T}, BioAlignments.PairwiseAlignment{S, S}}} where {S, T}"><code>translate</code></a> function. This could be useful if variants were called against a reference sequence for the entire species, but need to be analyzed as variants of a subtype later.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; ovis_human_alignment =
           PairwiseAlignment(AlignedSequence(human, Alignment(&quot;32M&quot;, 1, 1)), ovine)</code><code class="nohighlight hljs ansi" style="display:block;">BioAlignments.PairwiseAlignment{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}}:
  seq:  1 GACAGGCTGCATCAGAAGAGGCCATCAAGCAG 32
          ||| ||||||||  || |  | || ||| |||
  ref:  1 GACCGGCTGCATTCGAGGCTGTCAGCAAACAG 32</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; SequenceVariation.translate(bos_ovis_haplotype, ovis_human_alignment)</code><code class="nohighlight hljs ansi" style="display:block;">Haplotype{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}}, BioSymbols.DNA} with 2 edits:
  C22T
  G29A</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../variations/">Working with variations »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Saturday 28 January 2023 20:07">Saturday 28 January 2023</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
